### ISSUE TITLE
[Realtime] Ably-integration för Live Quiz (token-auth, teacher control, student answers)

### ISSUE THREAD

**Issue by pinkknives** [pinkknives/skolapp-v3#159]
Mål
Inför realtid för live-quiz: läraren styr flödet (start/next/end), elever skickar svar, och närvaro visas. Använd Ably med token-autentisering (server mintar kortlivade tokens). Inga hemliga nycklar i klienten.

⸻

Scope
	•	Installera och konfigurera ably.
	•	Server-endpoint som mintar token requests med capability-scoping per roll (teacher/student).
	•	Klienthelper för Ably (token-auth).
	•	Kanalstruktur:
	•	quiz:{quizId}:control – läraren publicerar start, next, end.
	•	quiz:{quizId}:answers – elever publicerar {questionId, answer}.
	•	quiz:{quizId}:room – presence (läraren/elever i rummet).
	•	Hooks/Actions:
	•	startQuiz, nextQuestion, endQuiz, submitAnswer, useQuizControl, joinRoom.
	•	UI-integration (minimal):
	•	Lärarvy: knappar för Start, Nästa, Avsluta + räknare för närvaro.
	•	Elevvy: lyssna på control, visa aktuell fråga, skicka svar.
	•	Read-only display (t.ex. skärm i klassrum): möjlighet att använda subscribe-only key som endast får ligga i publik miljö för rena konsumenter (ej inloggade användare).

⸻

Icke-mål (Out of scope)
	•	Beständig aggregering/poängräkning på servern.
	•	Ljud/animationer och avancerad moderering.
	•	Långtidsloggning och replay av historik.

⸻

Säkerhet
	•	Rotnyckel får aldrig till klienten.
	•	Token-TTL ≈ 30–60 min.
	•	Capabilities begränsar publish/subscribe/presence per kanal och roll.
	•	Rotera nycklar regelbundet.

⸻

Miljövariabler

/.env.example

# Server (hemligt)
ABLY_API_KEY=xxxx:yyyy
ABLY_TOKEN_TTL_SECONDS=3600

# Endast för read-only skärmar (valfritt)
NEXT_PUBLIC_ABLY_SUBSCRIBE_KEY=

Paket

npm i ably

Kodändringar (förslag på filstruktur)

1) Token-endpoint (Next.js App Router)

/app/api/ably/token/route.ts

import { NextRequest, NextResponse } from 'next/server';
import Ably from 'ably/promises';

function deriveRole(req: NextRequest): 'teacher' | 'student' {
  // TODO: byt till riktig auth/session. Temporärt: header/cookie eller fallback.
  const role = req.headers.get('x-role');
  return role === 'teacher' ? 'teacher' : 'student';
}

export async function GET(req: NextRequest) {
  const apiKey = process.env.ABLY_API_KEY!;
  const ttl = Number(process.env.ABLY_TOKEN_TTL_SECONDS ?? 3600);
  const role = deriveRole(req);
  const clientId = req.headers.get('x-client-id') ?? role;

  const ably = new Ably.Rest(apiKey);

  // Capability per roll: begränsa rättigheter
  // OBS: scopea specifikt quizId om ni har det i queryn; detta är en generell default.
  const capability =
    role === 'teacher'
      ? {
          'quiz:*:control': ['publish', 'subscribe'],
          'quiz:*:answers': ['subscribe'],
          'quiz:*:room': ['presence.subscribe', 'presence.enter', 'presence.leave'],
        }
      : {
          'quiz:*:control': ['subscribe'],
          'quiz:*:answers': ['publish'],
          'quiz:*:room': ['presence.subscribe', 'presence.enter', 'presence.leave'],
        };

  const tokenRequest = await ably.auth.createTokenRequest({
    clientId,
    ttl: ttl * 1000,
    capability: JSON.stringify(capability),
  });

  return NextResponse.json(tokenRequest);
}

) Ably-klient i webben

/src/lib/ablyClient.ts

import Ably from 'ably/promises';

let client: Ably.Realtime | null = null;

export function getAbly(clientId?: string, role?: 'teacher' | 'student') {
  if (!client) {
    client = new Ably.Realtime.Promise({
      authUrl: '/api/ably/token',
      authHeaders: {
        ...(clientId ? { 'x-client-id': clientId } : {}),
        ...(role ? { 'x-role': role } : {}),
      },
      transportParams: { remainPresentForMs: 60000 },
    });
  }
  return client;
}

3) Realtime-API (hooks/actions)

/src/lib/realtime/quiz.ts

import { getAbly } from '@/lib/ablyClient';

// Typning
export type ControlMsg =
  | { name: 'start'; data?: { at: number } }
  | { name: 'next';  data: { questionId: string } }
  | { name: 'end';   data?: {} };

export type AnswerMsg = { questionId: string; answer: string };

// Teacher actions
export async function startQuiz(quizId: string) {
  const ably = getAbly('teacher-1', 'teacher');
  const ch = ably.channels.get(`quiz:${quizId}:control`);
  await ch.publish('start', { at: Date.now() });
}
export async function nextQuestion(quizId: string, questionId: string) {
  const ably = getAbly('teacher-1', 'teacher');
  const ch = ably.channels.get(`quiz:${quizId}:control`);
  await ch.publish('next', { questionId });
}
export async function endQuiz(quizId: string) {
  const ably = getAbly('teacher-1', 'teacher');
  const ch = ably.channels.get(`quiz:${quizId}:control`);
  await ch.publish('end', {});
}

// Student actions
export async function submitAnswer(quizId: string, payload: AnswerMsg) {
  const ably = getAbly('student-1', 'student');
  const ch = ably.channels.get(`quiz:${quizId}:answers`);
  await ch.publish('answer', payload);
}

// Presence
export async function joinRoom(quizId: string, role: 'teacher'|'student', name: string) {
  const ably = getAbly(name, role);
  const room = ably.channels.get(`quiz:${quizId}:room`);
  await room.presence.enter({ role, name });
}

/src/hooks/useQuizControl.tsx

import { useEffect, useState } from 'react';
import { getAbly } from '@/lib/ablyClient';

type State = { phase: 'idle'|'running'|'ended'; questionId?: string };

export function useQuizControl(quizId: string): State {
  const [state, setState] = useState<State>({ phase: 'idle' });

  useEffect(() => {
    const ably = getAbly('student-1', 'student');
    const ch = ably.channels.get(`quiz:${quizId}:control`);
    ch.subscribe((msg) => {
      if (msg.name === 'start') setState({ phase: 'running' });
      if (msg.name === 'next')  setState({ phase: 'running', questionId: msg.data?.questionId });
      if (msg.name === 'end')   setState({ phase: 'ended' });
    });
    return () => { ch.detach(); };
  }, [quizId]);

  return state;
}

4) Minimal UI-koppling (exempel)
	•	Lärarsidan för ett quiz: lägg till knappar som anropar startQuiz, nextQuestion, endQuiz och en lista som visar presence från quiz:{id}:room.
	•	Elevsidan (join via kod/QR): anropa joinRoom, lyssna med useQuizControl, rendera aktuell fråga, skicka svar med submitAnswer.

5) Read-only display (valfritt)

För publik visning (enbart prenumerera):

import Ably from 'ably/promises';
const client = new Ably.Realtime.Promise({
  key: process.env.NEXT_PUBLIC_ABLY_SUBSCRIBE_KEY, // subscribe only
});

Testplan
	1.	Starta appen i två olika webbläsarfönster.
	2.	Fönster A (läraren):
	•	Öppna quizvy, klicka Start → kontrollmeddelande start skickas.
	•	Klicka Nästa med questionId → elevens vy visar frågan.
	3.	Fönster B (elev):
	•	useQuizControl fångar start/next.
	•	Skicka svar → lärarens vy kan logga inkommande svar (konsol/logg).
	4.	Presence:
	•	Båda fönsterna visas i närvarolistan, lämna ett fönster → närvaro uppdateras.
	5.	Säkerhet:
	•	Verifiera att elev inte kan publicera på control (capability-fel).
	•	Verifiera att lärare inte publicerar på answers med fel roll (om så önskas).

⸻

Acceptanskriterier
	•	✅ Token-auth används; inga hemliga nycklar i klienten.
	•	✅ Läraren kan starta, stega och avsluta quiz och eleverna ser uppdateringar i realtid.
	•	✅ Elever kan skicka svar; meddelanden landar i quiz:{id}:answers.
	•	✅ Närvaro fungerar (enter/leave) och visas i UI.
	•	✅ .env.example uppdaterad och README beskriver hur man får igång realtid lokalt.
	•	✅ Capabilities begränsar rättigheter per roll.

⸻

Dokumentation
	•	Uppdatera README.md med:
	•	hur man skapar/roterar Ably-nycklar,
	•	hur token-endpointen fungerar,
	•	hur man sätter x-role/x-client-id tills riktig auth finns.

⸻

Rollback
	•	Feature flagga realtidsläget i UI (t.ex. NEXT_PUBLIC_FEATURE_LIVE_QUIZ).
	•	Vid problem: inaktivera knapparna och fall back till självtempo.

⸻

Notera: Rotera den exponerade “Root”-nyckeln i Ably-konsolen innan merge om den tidigare visats publikt.
